src/drv_vrtek/packet.c

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: BSL-1.0
+/*
+ * OpenHMD - Free and Open Source API and drivers for immersive technology.
+ */
+
+/* VR-Tek Driver */
+
+#include <stdio.h>
+#include <string.h>
+#include "vrtek.h"
+
+#ifdef _MSC_VER
+#define inline __inline
+#endif
+
+inline static uint8_t read_uint8(const uint8_t** buffer)
+{
+    uint8_t ret = **buffer;
+    *buffer += 1;
+    return ret;
+}
+
+inline static int16_t read_int16_2bu(const uint8_t** buffer)
+{
+    int16_t ret = **buffer | (*(*buffer + 1) << 8);
+    *buffer += 2;
+    return ret;
+}
+
+inline static int16_t read_int16_2bs(const uint8_t** buffer)
+{
+    const int8_t** buffer_s = (const int8_t**) buffer;
+    int16_t ret = **buffer_s | (*(*buffer_s + 1) << 8);
+    *buffer += 2;
+    return ret;
+}
+
+inline static void write_uint8(uint8_t** buffer, uint8_t val)
+{
+    **buffer = val;
+    *buffer += 1;
+}
+
+int vrtek_encode_command_packet(uint8_t command_num, uint8_t command_arg,
+                                uint8_t* data_buffer, uint8_t data_length,
+                                uint8_t* comm_packet)
+{
+    uint8_t* buffer = comm_packet;
+    write_uint8(&buffer, VRTEK_REPORT_CONTROL_OUTPUT);
+    write_uint8(&buffer, command_num);
+    write_uint8(&buffer, command_arg);
+    write_uint8(&buffer, data_length);
+
+    if (data_length) {
+        memcpy(buffer, data_buffer, data_length);
+        buffer += data_length;
+    }
+
+    uint8_t parity_byte = *comm_packet;
+    for (int i = 1; i < (data_length + 4); ++i) {
+        parity_byte ^= *(comm_packet + i);
+    }
+    write_uint8(&buffer, parity_byte);
+
+    /* return the total number of bytes in the command packet */
+    return (buffer - comm_packet);
+}
+
+int vrtek_decode_command_packet(const uint8_t* comm_packet,
+                                uint8_t* command_num, uint8_t* command_arg,
+                                uint8_t* data_buffer, uint8_t* data_length)
+{
+    const uint8_t* buffer = comm_packet;
+    uint8_t report_num = read_uint8(&buffer);
+    if (report_num != VRTEK_REPORT_CONTROL_INPUT) {
+        LOGE("received invalid report number for command packet response "
+             "(expected %d but got %d)",
+             VRTEK_REPORT_CONTROL_INPUT, report_num);
+        return -1;
+    }
+
+    *command_num = read_uint8(&buffer);
+    *command_arg = read_uint8(&buffer);
+    *data_length = read_uint8(&buffer);
+
+    if (*data_length) {
+        memcpy(data_buffer, buffer, *data_length);
+        buffer += *data_length;
+    }
+
+    uint8_t checksum = *(comm_packet + 1);
+    for (int i = 2; i < (*data_length + 4 + 1); ++i) {
+        checksum ^= *(comm_packet + i);
+    }
+
+    /* checksum should be zero if parity byte was correct */
+    if (checksum) {
+        return -2;
+    }
+
+    return 0;
+}
+
+static inline void hmd_quat_rotate(float* out, float* in)
+{
+    double pos_val = 0.707099999999999950794915548613;
+    double neg_val = -0.707099999999999950794915548613;
```

Comment:
That's a lot of digits for something very imprecise. Just make this 0.7071, or probably better, sin(M_PI/4).

Reply:
Very good point - I should definitely be using the constants. I just pushed a commit that does just that. Thanks!